http-auth/routes/routes_test.go

```diff
@@ -0,0 +1,166 @@
+		request, _ := http.NewRequest(http.MethodGet, "/", nil)
+		rr := httptest.NewRecorder()
+
+		handler := http.HandlerFunc(IndexHandler)
+
+		handler.ServeHTTP(rr, request)
+
+		AssertStatusOK(t, rr)
+	})
+
+	t.Run("POST /", func(t *testing.T) {
+		bodyHTML := "<em>Hello Test</em>"
+		b := strings.NewReader(bodyHTML)
+		request, _ := http.NewRequest(http.MethodPost, "/", b)
+		request.Header.Add("Content-Type", "text/html")
+
+		rr := httptest.NewRecorder()
+
+		handler := http.HandlerFunc(IndexHandler)
+
+		handler.ServeHTTP(rr, request)
+
+		AssertStatusOK(t, rr)
+
+		AssertBodyContains(t, rr, bodyHTML)
+	})
+
+	t.Run("GET /200", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/200", nil)
+		rr := httptest.NewRecorder()
+
+		handler := http.HandlerFunc(Handle200)
+
+		handler.ServeHTTP(rr, request)
+
+		AssertStatusOK(t, rr)
+	})
+
+	t.Run("GET /500", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/500", nil)
+		rr := httptest.NewRecorder()
+
+		handler := http.HandlerFunc(Handle500)
+
+		handler.ServeHTTP(rr, request)
+
+		AssertStatusInternalServerError(t, rr)
+	})
+
+	t.Run("GET /authenticated", func(t *testing.T) {
+		t.Run("no authorisation provided", func(t *testing.T) {
+
+			request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+			rr := httptest.NewRecorder()
+
+			handler := http.HandlerFunc(HandleAuthenticated)
+
+			handler.ServeHTTP(rr, request)
+
+			AssertStatusUnauthorized(t, rr)
+		})
+
+		t.Run("invalid authorisation provided", func(t *testing.T) {
+
+			request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+			request.Header.Add("Authorization", "Basic invalid")
+			rr := httptest.NewRecorder()
+
+			handler := http.HandlerFunc(HandleAuthenticated)
+
+			handler.ServeHTTP(rr, request)
+
+			AssertStatusUnauthorized(t, rr)
+		})
+
+		t.Run("valid authorisation provided", func(t *testing.T) {
+
+			request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+			request.Header.Add("Authorization", "Basic dGVzdHVzZXI6c29tZXN0cm9uZ1BXRA==")
+			rr := httptest.NewRecorder()
+
+			handler := http.HandlerFunc(HandleAuthenticated)
+
+			handler.ServeHTTP(rr, request)
+
+			AssertStatusOK(t, rr)
+			AssertBodyContains(t, rr, "Hello, testuser")
+		})
+	})
+
+	t.Run("GET /limited", func(t *testing.T) {
+		t.Run("test over the limit", func(t *testing.T) {
+			request, _ := http.NewRequest(http.MethodGet, "/limited", nil)
+			handler := http.HandlerFunc(HandleRateLimit)
+			var responses []httptest.ResponseRecorder
+			for i := 0; i < 100; i++ {
+				rr := httptest.NewRecorder()
+				handler.ServeHTTP(rr, request)
+				responses = append(responses, *rr)
+			}
+
+			statusOK, statusTooManyRequests := 0, 0
+			for i := 0; i < 100; i++ {
+				if responses[i].Code == http.StatusOK {
+					statusOK++
+				} else if responses[i].Code == http.StatusTooManyRequests {
+					statusTooManyRequests++
+				}
+			}
+
+			if statusOK != 70 {
```

Comment:
These are some very specific assertions, which probably rely on your computer being fast. I can imagine a world where if your computer was really busy doing other things (or, more likely, the CI server was really busy doing other things), it might take more than 1 second to make 100 requests, at which point the rate limiting would reset.

I'd suggest making your assertions a bit more relaxed, maybe that _at least_ 30 succeeded, that _at least_ 1 failed, and that the sum of succeeded and failed is the total number of requests.

Reply:
Makes sense,thanks!